code/datums/components/bayonet_charge.dm

```diff
@@ -0,0 +1,183 @@
+	var/slownesstime = 5
+	///the amount of time cooldown lasts in deciseconds
+	var/cooldowntime = 150
+
+	//Below are static variables that should not be changed, ever.
+	var/stepstaken = 0
+	var/chargedir = null
+
+	var/amcharging = FALSE
+	var/chargewhenmove = FALSE
+	var/incooldown = FALSE
+	var/obj/item/weapon/gun/weaponinhand
+
+	///represents the time when the stepstaken var resets
+	var/movementcontinuationlimit
+	///the guy that is charging
+	var/mob/living/carbon/human/charger
+	///the child action of this component
+	var/datum/action/bayonetcharge/child
+	///The list of melee/not gun weapons available to be charged with
+	var/list/enabled_melee_weapons = list(
+		/obj/item/weapon/powerfist,
+		/obj/item/weapon/claymore/mercsword/machete,
+		/obj/item/weapon/claymore/mercsword/officersword,
+		/obj/item/weapon/twohanded/spear/tactical,
+	)
+
+/datum/component/bayonetcharge/RemoveComponent()
+	. = ..()
+	child.remove_action(charger)
+
+/datum/component/bayonetcharge/Initialize(...)
+	. = ..()
+	weaponinhand = src.parent
+	charger = weaponinhand.gun_user
+	child = new
+	child.give_action(src.charger)
+	child.parent = src
+
+/datum/action/bayonetcharge
+	name = "Toggle Bayonet Charge"
+	action_icon_state = "charge_human"
+
+	var/datum/component/bayonetcharge/parent
+
+/datum/action/bayonetcharge/give_action(mob/M)
+	. = ..()
+
+/datum/action/bayonetcharge/remove_action(mob/M)
+
+
+	parent.charge_off()
+	. = ..()
+
+/datum/action/bayonetcharge/action_activate()
+	if(parent.amcharging)
+		return
+	if(parent.chargewhenmove == TRUE)
+		parent.charge_off()
+		return
+
+	parent.charge_on()
+
+/datum/component/bayonetcharge/proc/charge_on()
+	if(incooldown)
+		to_chat(charger, "<span class='warning'>Charge is still on cooldown!</span>")
+		return
+	chargewhenmove = TRUE
+	child.add_selected_frame()
+	 //just in case you didnt activate the action, charger would be null if you do.
+	to_chat(charger, "<span class='warning'>You stand ready to charge</span>")
+	RegisterSignal(charger, COMSIG_MOVABLE_MOVED, .proc/update_charging)
+	RegisterSignal(charger, COMSIG_ATOM_DIR_CHANGE, .proc/on_dir_change)
+
+/datum/component/bayonetcharge/proc/charge_off()
+	chargewhenmove = FALSE
+
+	to_chat(charger, "<span class='warning'>You no longer stand ready to charge</span>")
+	if(charger) //locks up if you dont check if charger is existant
+		UnregisterSignal(charger, COMSIG_MOVABLE_MOVED)
+		UnregisterSignal(charger, COMSIG_ATOM_DIR_CHANGE)
+		UnregisterSignal(charger, list(COMSIG_MOVABLE_PREBUMP_TURF, COMSIG_MOVABLE_PREBUMP_MOVABLE, COMSIG_MOVABLE_PREBUMP_EXIT_MOVABLE))
+	child.remove_selected_frame()
+	stop_charge()
+
+
+/datum/component/bayonetcharge/proc/on_dir_change(datum/source, old_dir, new_dir)
+	SIGNAL_HANDLER
+	if(old_dir == new_dir)
+		return
+	stop_charge()
+//called when the marine moves
+/datum/component/bayonetcharge/proc/update_charging(datum/source, atom/oldloc, direction, Forced)
+	SIGNAL_HANDLER_DOES_SLEEP
```

Comment:
well, that's probably that you are doing things wrong

Reply:
What?

Reply:
Signal handler should sleep is forbidden